mllib/src/main/scala/org/apache/spark/ml/tree/impl/LocalDecisionTree.scala

```diff
@@ -0,0 +1,575 @@
+      doneLearning = currentLevel + 1 >= metadata.maxDepth || estimatedRemainingActive == 0
+
+      if (!doneLearning) {
+        val bestSplits: Array[Option[Split]] = bestSplitsAndGains.map(_._1)
+
+        // Aggregate bit vector (1 bit/instance) indicating whether each instance goes left/right
+        val aggBitVector: RoaringBitmap = LocalDecisionTreeUtils.aggregateBitVector(partitionInfo,
+          bestSplits, numRows, splits)
+
+        // Create a copy of our bitvector
+        val bv = new BitSet(numRows)
+        val iter = aggBitVector.getIntIterator
+        while(iter.hasNext) {
+          bv.set(iter.next)
+        }
+
+        // Obtain a new partitionInfo instance describing our current training status; the offsets
+        // of each node in our arrays of columns
+        partitionInfo = partitionInfo.update(bv, numNodeOffsets, labels, metadata)
+      }
+      currentLevel += 1
+    }
+
+    // Done with learning
+    rootNode.toNode
+  }
+
+  /**
+   * Find the best splits for all active nodes.
+   *  - For each feature, select the best split for each node.
+   *
+   * @return  Array over active nodes of (best split, impurity stats for split),
+   *          where the split is None if no useful split exists
+   */
+  private[impl] def computeBestSplits(
+      partitionInfo: PartitionInfo,
+      labels: Array[Double],
+      metadata: DecisionTreeMetadata,
+      splits: Array[Array[Split]]): Array[(Option[Split], ImpurityStats)] = {
+    // For each feature, select the best split for each node.
+    // This will use:
+    //  - labels (the labels column)
+    // Returns:
+    //   for each active node, best split + info gain,
+    //     where the best split is None if no useful split exists
+
+    partitionInfo match {
+      case PartitionInfo(columns: Array[FeatureVector], nodeOffsets: Array[Int],
+        activeNodes: BitSet, fullImpurityAggs: Array[ImpurityAggregatorSingle]) => {
+
+        // Iterate over the active nodes in the current level.
+        val toReturn = new Array[(Option[Split], ImpurityStats)](activeNodes.cardinality())
+        val iter: Iterator[Int] = activeNodes.iterator
+        var i = 0
+        while (iter.hasNext) {
+          // Our iterator iterates left-to-right across the active node periphery
+          val nodeIndexInLevel = iter.next
+          // Features for the current node start at fromOffset and end at toOffset
+          val fromOffset = nodeOffsets(nodeIndexInLevel)
+          val toOffset = nodeOffsets(nodeIndexInLevel + 1)
+          // Get the impurity aggregator for the current node
+          val fullImpurityAgg = fullImpurityAggs(nodeIndexInLevel)
+          // Get the best split for each feature for the current node
+
+          // TODO(smurching): In PartitionInfo, keep track of which features are associated
+          // with which nodes and subsample here
+          val splitsAndStats =
+            columns.map { col =>
```

Comment:
Maybe pass in the node here, and actually split it, returning the new child nodes.  No need to return stats since they are stored in the split node already.  Probably.